sycamore/transforms/merge_elements.py

```diff
@@ -0,0 +1,130 @@
+from abc import ABC, abstractmethod
+
+from ray.data import Dataset
+
+from sycamore.data import Document, Element, BoundingBox
+from sycamore.plan_nodes import NonCPUUser, NonGPUUser, Transform, Node
+from sycamore.transforms.map import generate_map_function
+
+from typing import Tuple
+
+
+
+class ElementMerger(ABC):
+    @abstractmethod
+    def merge_elements(self, document: Document) -> Document:
+        pass
+
+class GreedyElementMerger(ElementMerger):
+
+    def __init__(
+        self,
+        tokenizer,
+        max_tokens
+    ):
+        self.tokenizer = tokenizer
+        self.max_tokens = max_tokens
+
+
+    def _should_merge(self, element1: Tuple[Element, int], element2: Tuple[Element, int]) -> bool:
+        if element2[0].type == 'Title':
+            return False
+        if element1[1] + 1 + element2[1] > self.max_tokens:
+            return False
+        return True
+
+    def _merge(self, element1: Tuple[Element, int], element2: Tuple[Element, int]) -> Tuple[Element, int]:
+        """
+        Merge two elements; the new element's fields will be set as:
+        
+            type: "Section"
+            binary_representation: elt1.binary_representation + elt2.binary_representation
+            text_representation: elt1.text_representation + '\n' + elt2.text_representation
+            bbox: the minimal bbox that contains both elt1's and elt2's bboxes
+            properties: elt1's properties + any of elt2's properties that are not in elt1
+                > note: if elt1 and elt2 have different values for the same property, we take elt1's value
+            
+            > note: if any input field is None we take the other element's field without merge logic
+
+        Args:
+            element1 (Tuple[Element, int]): the first element (and number of tokens in it)
+            element2 (Tuple[Element, int]): the second element (and number of tokens in it)
+
+        Returns:
+            Tuple[Element, int]: a new merged element from the inputs (and number of tokens in it)
+        """
+        elt1, tok1 = element1
+        elt2, tok2 = element2
+        new_toks = 0
+        new_elt = Element()
+        new_elt.type = "Section"
```

Comment:
How are we handling table (or other typed) elements?

Reply:
tables special data is stored in properties, so that gets persisted. If two tables get merged then this implementation will drop the second table's "rows" and "columns" values. But otherwise 'merged' elements are just basic Elements

Reply:
I feel we should not merge two different types of elements? e.g. author with title, or table with figure etc?

Reply:
I guess it sort of depends on how we're handling types? My impression was that types are essentially inherited directly from unstructured, in which case we want to be able to merge 'Title' and 'NarrativeText' - that's the whole point. Basically we want to squash context together, and that can come from different sources. 

If we are gonna make our type system more uniform then it might make sense to filter only the 'Text' elements for this transform or something, but as it is I think anything with a text_representation can and should be merged (given sufficient room in the context window)